### PR TITLE
Allow `-preview=in` only with `extern(D|C++)`

### PR DESCRIPTION
**Pull request by Geod24**
```
The intent of `-preview=in` is to make `in` the go-to storage class for input parameters.
In doing so, it makes the ABI less obvious (although still deterministic),
by sometimes degrading the parameter from reference to a value if it is better ABI-wise.
This behavior is obviously at odds with non-`extern(D)` functions,
which expect to match a specific ABI, hence `in` parameters can no longer apply
to non-`extern(D)` or `extern(C++)` functions when `-preview=in` is used.

However, as C++, also have a "go to" storage class for input parameters (`const T&`),
`in` can also be applied on `extern(C++)` function in order to bind to `const T&` parameters.
When doing so, `in` will always mean `ref`. This also allows to expose a closer API
for a function than via `const ref`, as `in` will allow to bind rvalues to `const T&`, as in C++.
```

@kinke @ibuclaw : Any opinion on this ? If this is too much, I can see two other course of action:
- Disabling `in` for anything non-`extern(D)`;
- Making `in` always make `scope const` with `extern(C)`, never `ref`, disable for others;

### CODE REVIEW

**Comment by Geod24** [comment]
Maybe the `const` solution for `extern(C)` is the best way forward. After all, we allow to pass D arrays to `extern(C)` functions (https://github.com/dlang/druntime/blob/0125219f894a43d2ab0a7571b690c96e400051ca/src/rt/aApply.d#L21).

**Comment by PetarKirov** [comment]
Keep in mind that `scope` is meaningless for non-`extern (D)` functions that are not implemented in D, as C and C++ compilers won't enforce that the parameter are not escaped, so while `in` means `const scope maybe-ref` in D, it should be translated as exactly `const ref` for C and C++.

**Comment by kinke** [comment]
> Disabling `in` for anything non-`extern(D)`

I think I'd prefer that for simplicity, otherwise the `in` semantics get too complicated.

**Comment by Geod24** [comment]
> Keep in mind that `scope` is meaningless for non-`extern (D)` functions that are not implemented in D

Not exactly meaningless, since the compiler will check `scope` if called from a `@safe` context. And yes, we have some in druntime.

> I think I'd prefer that for simplicity, otherwise the `in` semantics get too complicated.
> [...]
> Hmm okay that'd be nice indeed...

That was exactly my thought process. Maybe @ibuclaw can weight in ?

**Comment by PetarKirov** [comment]
> compiler will check scope if called from a @safe context

That's actual the main problem. `scope` parameters allow  allow more code in `@safe` functions, not less. For example, functions can to receive pointers/references to local variables in `@safe` code via scope parameters. This is under the assumption that the callee is implemented in D and the compiler has verified that it doesn't escape the address of its scope parameters. If the function is implemented in another language, this assumption doesn't hold any more.

**Comment by Geod24** [comment]
> If the function is implemented in another language, this assumption doesn't hold any more.

The same can be said of other attributes. IMO C++ `const`'s situation is even worse, are we could be forced to apply `const` to a parameter even though it is mutated.

**Comment by atilaneves** [comment]
I think it makes sense to be able to declare parameters to be `scope` even in `extern(C)` code, especially considering that just because functions have that linkage doesn't imply they're actually written in C.

And even if they are, I for one would like to tell the compiler that they're not escaped.

**Comment by anon17** [comment]
Is this temporary? If it wasn't always `ref` per recent plan, `in` with intended optimizations would be the goto storage class for C functions.

Last I checked `const T&` wasn't the default storage class for input parameters. I think `const T&` is better written as `in ref T`. Will it then accept rvalues?

**Comment by Geod24** [comment]
I think this is finally passing the CI (!)

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
Final ping: @kinke @ibuclaw

File: changelog/previewInLink.dd [hunk 1]
```diff
@@ -0,0 +1,11 @@
+`-preview=in` can now be used with `extern(C++)`, disabled for other non-D linkage
+
+The intent of `-preview=in` is to make `in` the go-to storage class for input parameters in D.
```
[the remark this answers is not in the document]
Comment (Geod24) [reply changelog/previewInLink.dd:3]: Has Walter said those words exactly? Not that I recall. But he did put `in`, `inout` (now `ref`) and `out` in the language, so yeah I think that's as close as we'd get.

File: changelog/previewInLink.dd [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+However it is D centric, as it is an enhanced version of `scope const ref`.
+As non-`extern(D)` functions usually are expected to match a specific ABI,
+using `in` is hardly a good idea.
+
+However, as C++, also have a "go to" storage class for input parameters (`const T&`),
```
[the remark this answers is not in the document]
Comment (Geod24) [reply changelog/previewInLink.dd]: Fixed (also dropped the `However` as it was repetitive with a previous line

**Comment by Geod24** [comment]
@dkorpel : Fixed

**Comment by ibuclaw** [comment]
I have nothing against this, would the likes of @atilaneves or @WalterBright wish to have another look?

**Review by ibuclaw — APPROVED** [review]

**Review by kinke — APPROVED** [review]

File: src/dmd/typesem.d [hunk 3]
```diff
@@ -1224,6 +1224,25 @@ extern(C++) Type typeSemantic(Type type, const ref Loc loc, Scope* sc)
...
+                // -preview=in: Always add `ref` when used with `extern(C++)` functions
+                // Done here to allow passing opaque types with `in`
+                if (global.params.previewIn && (fparam.storageClass & (STC.in_ | STC.ref_)) == STC.in_)
+                {
+                    switch (tf.linkage)
+                    {
+                    case LINK.cpp:
+                        fparam.storageClass |= STC.ref_;
+                        break;
+                    case LINK.default_, LINK.d:
```
Comment (WalterBright) [inline src/dmd/typesem.d:1236]: What exactly does `in` mean for D linkage? Is it different than for C++?

**Comment by WalterBright** [comment]
Does the meaning of `in` differ for D and C++ linkage?

File: src/dmd/typesem.d (hunk 3 above)
Comment (Geod24) [reply src/dmd/typesem.d:1236]: This PR predates our decision to make `in` always mean `ref`. I'll correct it in the next PR.

**Comment by atilaneves** [comment]
> Does the meaning of `in` differ for D and C++ linkage?

I hope not.

**Comment by Geod24** [comment]
Not after I disable the optimization, which is the next PR.

**Pull request merged**